### PR TITLE
Fix lister generation so it's only for api groups

### PR DESCRIPTION
**Pull request by ncdc**
cc @mfojtik @smarterclayton @deads2k @liggitt @sttts @soltysh @pweil- @bparees @derekwaynecarr @pmorie

### CODE REVIEW

**Comment by ncdc** [comment]
FYI, the "core" in the package paths (e.g. pkg/user/client/listers/core/...) comes from the fact that all of our apis have "api" in their package path and not the name of the group. For example, in kube, types for the authentication group are in pkg/apis/authentication and pkg/apis/authentication/v1. The tooling sees `authentication` in the path and uses that as the group name (pkg/client/listers/authentication, in kube upstream). But because we have pkg/user/api and not pkg/apis/user, the tooling translates "api" to "core" (because in kube, pkg/api is the core api).

File: pkg/authorization/client/listers/core/internalversion/expansion_generated.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// This file was automatically generated by lister-gen
+
+package internalversion
+
+// PolicyListerExpansion allows custom methods to be added to
+// PolicyLister.
+type PolicyListerExpansion interface{}
+
+// PolicyNamespaceListerExpansion allows custom methods to be added to
+// PolicyNamespaeLister.
```
Comment (stevekuznetsov) [inline pkg/authorization/client/listers/core/internalversion/expansion_generated.go]: `s/Namespae/Namespace/` -- got copy-pasta'd a couple times

**Comment by deads2k** [comment]
We've talked about moving all the API types to match kube patterns to ease generation like this.  Do we want to bite the bullet now?

Even though its stuttery, I would organize them the areas like mini-repos and do something like `pkg/deploy/apis/deploy/v1`.

**Comment by ncdc** [comment]
I saw you did a hack for client-gen. We could do that for listers and
informers if we don't want to move packages around.

**Comment by deads2k** [comment]
> I saw you did a hack for client-gen. We could do that for listers and
> informers if we don't want to move packages around.

I think we'll be chasing hacks forever if don't switch our type packages to be kube-like.

**Comment by deads2k** [comment]
> I think we'll be chasing hacks forever if don't switch our type packages to be kube-like.

I did that because the alternative being proposed (double down on hand created clients) was worse than the hack I put in.

**Comment by ncdc** [comment]
I'm happy to do the move

**Comment by smarterclayton** [comment]
Shouldn't the short name come from the comment in the API group?  Why are
paths special?

**Comment by deads2k** [comment]
> Shouldn't the short name come from the comment in the API group?  Why are
> paths special?

The generators like getting an APIs package then make assumptions about the shape within.  We could try to swim against the tide and claim (correctly) that the generator shouldn't care about the shape of the packages, but ultimately that seems like wasted effort.  The more we look like kube, the easier it is to switch back and forth.

**Comment by ncdc** [comment]
The only generator that currently uses the groupName comment is client-gen. We could extend the others to use it, or we could just go by path, which works for all the apis in the kube repo, service catalog, and most likely 100% of any other extension apiserver repos that are out there.

To punt on this decision in the very short term, I'm going to try to pull in the same hack David did for client-gen over to the listers and informers so we don't have to change anything else for the time being.

**Comment by ncdc** [comment]
Replaced by #13982

**Pull request closed without merging**